File: sliding_window/best_time_to_buy_and_sell_stock.py
# 121: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buyingPrice = prices[0]
        maxProfitMade = 0

        for day in range(1, len(prices)):
            price = prices[day]
            if price < buyingPrice:
                buyingPrice = price
            
            maxProfitMade = max(maxProfitMade, price - buyingPrice)

        return maxProfitMade


c = Solution()

prices1 = [7,1,5,3,6,4]
prices2 = [7,6,4,3,1]

print(c.maxProfit(prices1))
print(c.maxProfit(prices2))

File: sliding_window/best_time_to_buy_and_sell_stock.js
// 121: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let max = 0,
    buyingPrice = prices[0];

  for (let i = 1; i < prices.length; i++) {
    if (prices[i] < buyingPrice) {
      buyingPrice = prices[i];
    }

    const profit = prices[i] - buyingPrice;

    max = Math.max(max, profit);
  }

  return max;
};
